DifferenceGenerator/src/main/kotlin/algorithms/DivideAndConquerAligner.kt

```diff
@@ -0,0 +1,155 @@
+        // get hashes and mark duplicates
+        hashes1 = markDuplicates(hasher.getHashes(a))
+        hashes2 = markDuplicates(hasher.getHashes(b))
+
+        // find unique exact matches between the two sequences
+        val equals = findMatches()
+
+        var lastMatchIndex1 = 0
+        var lastMatchIndex2 = 0
+
+        // reset the sequence iterators to the start
+        a.reset()
+        b.reset()
+
+        for (matchPair in equals) {
+            // create sub-sequences up to the next match to be aligned
+            val subArray1 = a.take(matchPair.first - lastMatchIndex1)
+            val subArray2 = b.take(matchPair.second - lastMatchIndex2)
+
+            // create alignment up to the next match
+            alignment.addAll(getSubAlignment(subArray1, subArray2))
+
+            // insert a match for the currently selected exact match
+            alignment += AlignmentElement.MATCH
+
+            // update last match indices
+            lastMatchIndex1 = matchPair.first + 1
+            lastMatchIndex2 = matchPair.second + 1
+
+            // advance iterators to skip the current match's positions
+            a.next()
+            b.next()
+        }
+
+        // process alignment after last known match
+        val subArray1 = a.take(a.size() - lastMatchIndex1)
+        val subArray2 = b.take(b.size() - lastMatchIndex2)
+        alignment.addAll(getSubAlignment(subArray1, subArray2))
+
+        return alignment.toTypedArray()
+    }
+
+    /**
+     * Aligns two sequences by the given alignment algorithm.
+     *
+     * For performance reasons, we use shortcuts if one of the sequences is empty.
+     */
+    private fun getSubAlignment(
+        slice1: List<T>,
+        slice2: List<T>,
+    ): Array<AlignmentElement> {
+        return if (slice1.isEmpty() && slice2.isEmpty()) {
+            arrayOf()
+        } else if (slice1.isEmpty()) {
+            Array(slice2.size) { AlignmentElement.INSERTION }
+        } else if (slice2.isEmpty()) {
+            Array(slice1.size) { AlignmentElement.DELETION }
+        } else {
+            // gather alignment between two exact matches (or start and first match)
+            // this runs the underlying alignment algorithm aka this is the divide and conquer step
+            algorithm.run(ArrayList(slice1), ArrayList(slice2))
+        }
+    }
+
+    /**
+     * Marks duplicates in the given array of hashes.
+     *
+     * @param hashArray The array of hashes to mark duplicates in.
+     * @return The array of hashes with duplicates marked as empty byte-arrays.
+     */
+    private fun markDuplicates(hashArray: Array<ByteArray>): Array<ByteArray> {
+        val duplicates: Set<Int> = setOf()
+        for (i in hashArray.indices) {
+            for (j in i + 1 until hashArray.size) {
+                // if two hashes are equal, put the indices of both in the duplicates set
+                if (hashArray[i].contentEquals(hashArray[j])) {
+                    duplicates.plus(j)
```

Comment:
`.plus()` creates a new set with includes the added element, but the new set is never used. Is it intentional to do this?
I thought that either `.add(j)` or `duplicates = duplicates.plus(j)` would be the right way but maybe I am wrong.

Reply:
The tests currently dont allow duplicate images in one video, so yeah, this wasnt caught...